File: backend/src/modules/audit/audit.service.ts
import { AuditLog } from '@database/index';
import { AppError } from '@utils/AppError';
import { parsePagination } from '@utils/pagination';
import { Op, WhereOptions } from 'sequelize';

export async function listAuditLogs(filters: {
  page?: string;
  limit?: string;
  action?: string;
  module?: string;
  userId?: string;
  search?: string;
  sort?: string;
  order?: 'asc' | 'desc';
}) {
  const { page, limit, offset, sort, order } = parsePagination(filters);

  const where: WhereOptions<AuditLog> = {};
  if (filters.action) {
    where.action = filters.action;
  }
  if (filters.module) {
    where.module = filters.module;
  }
  if (filters.userId) {
    where.userId = Number(filters.userId);
  }
  if (filters.search) {
    where[Op.or as never] = [
      { action: { [Op.like]: `%${filters.search}%` } },
      { entityType: { [Op.like]: `%${filters.search}%` } },
    ];
  }

  const { rows, count } = await AuditLog.findAndCountAll({
    where,
    limit,
    offset,
    order: [[sort, order]],
    distinct: true,
  });

  return {
    logs: rows,
    meta: { page, limit, total: count, totalPages: Math.ceil(count / limit) },
  };
}

export async function getAuditLog(id: number) {
  const log = await AuditLog.findByPk(id);
  if (!log) {
    throw AppError.notFound('Audit log not found');
  }
  return log;
}

File: backend/app/services/audit_service.py
from app.core.errors import AppError
from app.core.pagination import order_by_clause, parse_pagination
from app.models import AuditLog


def list_audit_logs(db, filters):
    page, limit, offset, sort, order = parse_pagination(filters)

    q = db.query(AuditLog)
    if filters.get('action'):
        q = q.filter(AuditLog.action == filters['action'])
    if filters.get('module'):
        q = q.filter(AuditLog.module == filters['module'])
    if filters.get('userId'):
        q = q.filter(AuditLog.userId == int(filters['userId']))
    if filters.get('search'):
        term = '%{}%'.format(filters['search'])
        q = q.filter(db.or_(AuditLog.action.like(term), AuditLog.entityType.like(term)))

    total = q.count()
    logs = q.order_by(order_by_clause(AuditLog, sort, order)).offset(offset).limit(limit).all()

    return {
        'items': logs,
        'meta': {'page': page, 'limit': limit, 'total': total, 'totalPages': (total + limit - 1) // limit},
    }


def get_audit_log(db, log_id):
    log = db.query(AuditLog).filter(AuditLog.id == log_id).first()
    if not log:
        raise AppError.not_found('Audit log not found')
    return log